src/tokenizer.rs

```diff
@@ -0,0 +1,510 @@
+#![allow(unused_variables, dead_code)]
+
+use std::{mem, vec};
+
+#[derive(Copy, Clone, Debug)]
+#[repr(u8)]
+pub enum TokenType {
+    Ampersand = b'&',
+    Asterisk = b'*',
+    At = b'@',
+    Comma = b',',
+    Colon = b':',
+    Semicolon = b';',
+    OpenParenthesis = b'(',
+    CloseParenthesis = b')',
+    OpenSquare = b'[',
+    CloseSquare = b']',
+    Dollar = b'$',
+    Tilde = b'~',
+    Caret = b'^',
+    Plus = b'+',
+    Equals = b'=',
+    Pipe = b'|',
+    Greater = b'>',
+    Space = b' ',
+    SingleQuote = b'\'',
+    DoubleQuote = b'"',
+    Slash = b'/',
+    Bang = b'!',
+
+    Backslash = b'\\',
+    Cr = b'\r',
+    Feed = 0x0C, // \f
+    Newline = b'\n',
+    Tab = b'\t',
+
+    Comment = u8::MAX,
+    Word = u8::MAX - 1,
+    Combinator = u8::MAX - 2,
+}
+
+impl From<u8> for TokenType {
+    fn from(ch: u8) -> Self {
+        assert!(is_char_valid_token(ch));
+        // SAFETY: Always valid 'cause we assert it above
+        unsafe { mem::transmute(ch) }
+    }
+}
+
+impl From<TokenType> for u8 {
+    fn from(ty: TokenType) -> Self {
+        // SAFETY: Always valid 'cause it's memory repr is u8
+        unsafe { mem::transmute(ty) }
+    }
+}
+
+const fn is_char_valid_token(ch: u8) -> bool {
+    match ch {
+        _ if ch == b'&' => true,
+        _ if ch == b'*' => true,
+        _ if ch == b'@' => true,
+        _ if ch == b',' => true,
+        _ if ch == b':' => true,
+        _ if ch == b';' => true,
+        _ if ch == b'(' => true,
+        _ if ch == b')' => true,
+        _ if ch == b'[' => true,
+        _ if ch == b']' => true,
+        _ if ch == b'$' => true,
+        _ if ch == b'~' => true,
+        _ if ch == b'^' => true,
+        _ if ch == b'+' => true,
+        _ if ch == b'=' => true,
+        _ if ch == b'|' => true,
+        _ if ch == b'>' => true,
+        _ if ch == b' ' => true,
+        _ if ch == b'\'' => true,
+        _ if ch == b'"' => true,
+        _ if ch == b'/' => true,
+        _ if ch == b'!' => true,
+
+        _ if ch == b'\\' => true,
+        _ if ch == b'\r' => true,
+        _ if ch == 0x0C => true, // \f
+        _ if ch == b'\n' => true,
+        _ if ch == b'\t' => true,
+
+        // TODO(CGQAQ): Maybe not return true?
+        _ if ch == u8::MAX => true,
+        _ if ch == u8::MAX - 1 => true,
+        _ if ch == u8::MAX - 2 => true,
+
+        _ => false,
+    }
+}
+
+///
+/// Check if a char is not allowed to escape
+/// * return true if it's not allowed to escape.
+///
+#[inline(always)]
```

Comment:
should not add inline always , let compiler optimize it for you or add it after bench.